sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -348,26 +348,76 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
         Some(sources.EqualTo(a.name, v))
       case expressions.EqualTo(Literal(v, _), a: Attribute) =>
         Some(sources.EqualTo(a.name, v))
+      case expressions.EqualTo(Cast(a: Attribute, _), Literal(v, _))
+        if a.dataType == DateType || a.dataType == TimestampType =>
+          Some(sources.EqualTo(a.name, v))
```

Comment:
when we are comparing DateType column with a String literal, we could not cast the String literal value to DateType since it would `eval` to the inner representation of `DateType`, i.e. int, which means nothing for JDBC source.